clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -3845,6 +3716,9 @@ void handleResponse(AbstractResponse abstractResponse) {
 
                         KafkaFutureImpl<ListOffsetsResultInfo> future = futures.get(result.getKey());
                         PartitionData partitionData = result.getValue();
+                        Errors error = partitionData.error;
+                        if (context.shouldRefreshMetadata(error))
+                            rescheduleMetadataTask(context, () -> calls);
```

Comment:
It's a little weird to retry all of the calls when just one partition may have had a metadata error, right? I think I'd suggest we try to make this a little simple for now. While iterating over the partitions in the response, we can collect the set of partitions with metadata errors. Then we can retry this call individually for just those partitions with metadata errors. Does that make sense?

Reply:
Yes we should definitively do that. Unfortunately I need to head off now to catch my plane. I'll take a look next week. 

Thanks for the reviews and the quick feedback!

Reply:
My concern here is that we are retrying all the calls, not just the one that failed. If there are N calls and each one fails, then will will retry N*N requests. The simple approach is just to retry this call only. What I am thinking is something like this:

```java
        for (final Map.Entry<Node, Map<TopicPartition, ListOffsetRequest.PartitionData>> entry: leaders.entrySet()) {
            final int brokerId = entry.getKey().id();
            final Map<TopicPartition, ListOffsetRequest.PartitionData> partitionsToQuery = entry.getValue();

            calls.add(new Call("listOffsets on broker " + brokerId, context.deadline(), new ConstantNodeIdProvider(brokerId)) {

                @Override
                ListOffsetRequest.Builder createRequest(int timeoutMs) {
                    return ListOffsetRequest.Builder
                            .forConsumer(true, context.options().isolationLevel())
                            .setTargetTimes(partitionsToQuery);
                }

                @Override
                void handleResponse(AbstractResponse abstractResponse) {
                    ListOffsetResponse response = (ListOffsetResponse) abstractResponse;
                    Set<TopicPartition> partitionsWithErrors = new HashSet<>();

                    for (Entry<TopicPartition, PartitionData> result : response.responseData().entrySet()) {
                        KafkaFutureImpl<ListOffsetsResultInfo> future = futures.get(result.getKey());
                        PartitionData partitionData = result.getValue();
                        Errors error = partitionData.error;
                        if (context.shouldRefreshMetadata(error)) {
                            partitionsWithErrors.add(result.getKey())
                        } else if (partitionData.error == Errors.NONE) {
                            future.complete(new ListOffsetsResultInfo(partitionData.offset, partitionData.timestamp, partitionData.leaderEpoch));
                        } else {
                            future.completeExceptionally(result.getValue().error.exception());
                        }
                    }

                    if (!partitionsWithErrors.isEmpty()) {
                        partitionsToQuery.keySet().retainAll(partitionsWithErrors);
                        rescheduleMetadataTask(context, () -> Collections.singletonList(this));
                    }
                }

                @Override
                void handleFailure(Throwable throwable) {
                    for (TopicPartition tp : partitionsToQuery.keySet()) {
                        KafkaFutureImpl<ListOffsetsResultInfo> future = futures.get(tp);
                        future.completeExceptionally(throwable);
                    }
                }
            });
```

The important part is `rescheduleMetadataTask(context, () -> Collections.singletonList(this));`. We only reschedule calls individually. And even better would be to just create a separate operation context for each retried call so that there are no concurrency problems.